### ISSUE TITLE
Cart tests after refactor

### ISSUE THREAD

**Issue by marlass** [SAP/cloud-commerce-spartacus-storefront#5051]
Cart refactor have a lot of changes in our logic. We need to test it to make sure everything that worked correctly before still works and that a newly added behavior also works as intended.

Use cases that might help:

- adding to cart now supports multiple calls (you should be able to add multiple products before the cart was created and all should end up in cart)
- cart merging for anonymous user (anonymous cart merged to user cart)
- cart merging for guest user (logged user cart and guest moved to user cart)
- logged user should always have one cart (create cart should be triggered only when there is no current cart)
- updating entries should work for anonymous/logged/guest
- removing entries should word for anonymous/logged/guest
- the current cart should be loaded for the user after logging in (when anonymous cart exists - merge)
- after finishing order there should be an attempt to load current cart (only for logged user)
- after removing cart activeCart returns empty object
- cart is saved in browser for anonymous/logged user
- loading status returned correctly when cart loads
- loading is shown for whole adding to cart process (from clicking till item was added to cart - under the hood there is a load attempt and then create cart)
- entries shown correctly in all places (cart/mini cart)
- guest checkout still works as before
- cart getters return correct results (loading status, entry, assignedUser etc.)
- cart is refreshed after each update/add/remove
- no "Cart not found" errors
- cart is created only if there is no current cart
- cart should be removed from store after order
- cart state should be the same in `multi-cart` and `cart` branch
- the cart should be reloaded on currency/language change
- cart entries should be in order returned from backend
- checkout should still work without issues/errors
- validate cart with ASM

**Comment by KateChuen** [comment]
- [ ] adding to cart now supports multiple calls (you should be able to add multiple products before the cart was created and all should end up in cart)
- [x] cart merging for anonymous user (anonymous cart merged to user cart)
- [x] cart merging for guest user (logged user cart and guest moved to user cart)
- [x] logged user should always have one cart (create cart should be triggered only when there is no current cart) verified with postman call that we have only one cart, even with merged carts, or logged on different browsers
- [x] updating entries should work for anonymous/logged/guest
- [x] removing entries should work for anonymous/logged/guest
- [x] the current cart should be loaded for the user after logging in (when anonymous cart exists - merge)
- [ ] after finishing order there should be an attempt to load current cart (only for logged user) -> **don't see it**
- [x] after removing cart activeCart returns empty object (logged user, anonymous user, guest user)
- [x] cart is saved in browser for anonymous/logged user
- [x] loading status returned correctly when cart loads
- [x] loading is shown for whole adding to cart process (from clicking till item was added to cart - under the hood there is a load attempt and then create cart)
- [ ] entries shown correctly in all places (cart/mini cart) -> **issue with qty after anonymous cart merge**
- [x] guest checkout still works as before
- [x] cart getters return correct results (loading status, entry, assignedUser etc.)
- [x] cart is refreshed after each update/add/remove
- [ ] no "Cart not found" errors -> **Got "cart not found" error a couple of times, but cannot pinpoint the cause.**
- [x] cart is created only if there is no current cart
- [ ] cart should be removed from store after order
- [ ] cart state should be the same in multi-cart and cart branch
- [x] the cart should be reloaded on currency/language change
- [x] cart entries should be in order returned from backend
- [x] checkout should still work without issues/errors
- [ ] validate cart with ASM -> **Found a couple of issues when customer is logged in a different browser and has already started a cart. Or if agent started a cart before starting the customer session**
- [x] cart with express checkout, cart merge, checkout etc
- [x] cart on different browsers

**Comment by KateChuen** [comment]
Marcin said "Hey, I tested once again those cases which wasn't checked and haven't found any issues. Ticket is now on CR "

I will do another quick round of testing (smoke test) once ticket https://github.com/SAP/cloud-commerce-spartacus-storefront/issues/4432 is in QA.

**Comment by KateChuen** [comment]
Tested it again in dev3. 

 **Issues found**
- logged user adds a product and proceeds to checkout. Logs out during checkout. Anonymous user adds product to cart, proceeds to checkout, logs in. Go on cart page, update the quantity of the product the logged user had. I'm getting this error:
`{
`  "errors" : [ {
  `   "message" : "item 8796094239639 no longer valid (was removed): object no longer valid",
 `    "type" : "JaloObjectNoLongerValidError"
`  } ]
`}
--> Marcin says: 2 requests on the cart at the same time. Unfortunately final fix for it requires breaking changes, so final fix will be available in 2.0. checkout and cart are currently separated in storefront, but use the same backend entities and we can't reliably avoid 2 requests in parallel

- Logged user adds a product to cart, proceeds to checkout, logs out during checkout. Anonymous user adds a product to cart, proceeds to checkout, logins. Getting this error: 
<img width="784" alt="Screen Shot 2019-12-03 at 10 58 57 AM" src="https://user-images.githubusercontent.com/45409515/70067449-4aba9c00-15bc-11ea-83da-8dbe278f619b.png">
<img width="620" alt="Screen Shot 2019-12-03 at 11 01 57 AM" src="https://user-images.githubusercontent.com/45409515/70067488-5a39e500-15bc-11ea-9d15-a99d8cb307a8.png">

- During checkout, the shipping page keeps loading, #5632

**Comment by marlass** [comment]
Separate ticket for `cart not found` bug: https://github.com/SAP/cloud-commerce-spartacus-storefront/issues/5647

**Comment by KateChuen** [comment]
Another issue found with asm. ticket #5648 
And another issue with the cart when logged in on different browsers: #5652 
Couldn't checkout with a guest user because of issue #5632 
And for the "item no longer valid" issue, Marcin has it as future improvement to do in 2.0.

**Comment by KateChuen** [comment]
After Marcin's fix, 
- the ASM issue is gone. 
- and able to checkout with a guest user now. 

Marcin said the multiple browsers issue is on develop branch too, and will be done after cart refactoring. 

Only issue left is the "item no longer valid" error and occasional "cart not found" error that will be tackled later.
